### ISSUE TITLE
Git: Support git worktrees in workspace

### ISSUE THREAD

**Issue by colemickens** [microsoft/vscode#68038]
```
$ mkdir repro
$ cd repro
$ mkdir example1
$ cd example1
$ git init; echo "hello" > world.txt; git add world.txt; git commit -m "init";
$ git worktree add ../example1branchA
$ git worktree add ../example1branchB
$ cd ..

$ mkdir example2
$ cd example2
$ git init; echo "hello" > world.txt; git add world.txt; git commit -m "init";
$ cd ..
```
Scenario:
1. Open `repro` directory in VS Code
2. Go to Git tab.

**Expected:** "Source Control Providers" lists Git repos.
  * `example1`
  * `example1branchA`
  * `example1branchB`
  * `example2`

**Actual:** "Source Control Providers" lists Git repos.
 * `example1`
 * `example2`

I sometimes will have lots of branches checked out in parallel on huge repos that I can't afford to clone multiple times. I would also like to be able to use VS Code's source control functionality with these checkouts all the same.

Thanks! and thanks for Code!

**Comment by colemickens** [comment]
Thank you honorable vscodebot, but this feature request is not a duplicate of that issue.

**Comment by charltonstanley** [comment]
i just tested this out and it looks to be fixed now on v1.35.0 @colemickens is this still an issue for you?
![image](https://user-images.githubusercontent.com/22397984/63544231-475a4380-c4ea-11e9-886c-6503a46f3893.png)

**Comment by JojoS62** [comment]
In current version 1.40.2 I can see the multiple source control providers only when the root does not contain a .git and is initialized by the command 'Git: Initialize Repository'.
After closing and reopening VSCode, only the git for the project root is shown, subdirs with own .git are no longer listed.

**Comment by rajatvijay** [comment]
I am facing a similar issue, any updates here?

**Comment by rohit04saluja** [comment]
Is someone working on this? Need the worktree support

**Comment by halfdan** [comment]
What's missing here is the ability to switch between branches and/or create a new worktree when working in a bare repo. The branch listing works just fine but vscode doesn't automatically change the working directories when I select a branch that exists as a worktree.

**Comment by HybridEidolon** [comment]
The worktree support also doesn't work if the worktree is located in another path, e.g.:

```
Code/project/.git -- actual repo
Code/project -- primary 
Code/project2/.git -- worktree of Code/project
```

and opening project and project2 in separate VS Code workspaces, only project will see the repo, not project2.

**Comment by cgarrovillo** [comment]
As an example,  the Gitlens [repo](https://github.com/gitkraken/vscode-gitlens) might be a good place to start as they have this included in their extension but behind a paywall.

**Comment by jeffrey-aguilera** [comment]
> I don't know why this keeps getting comments, afaict vs code supports work trees just fine these days.

Not with devcontainers.

**Comment by intrnl** [comment]
I have a setup where I have a folder with a bare-cloned repository and have worktrees for each branches I work on, for some reason it thinks all of my branches are unpublished, which it isn't the case.

```
.git          // gitdir: ./.bare
.bare/        // bare cloned repository
staging/
production/
development/
feature-xxx/ 
bugfix-xxx/
```

**Comment by afa7789** [comment]
Yeah my DevContainner is not working with versioncontrol. ( Linux user )
![image](https://user-images.githubusercontent.com/26887703/176911302-79992dc5-e62f-418c-ad6d-917f75e604b2.png)
![image](https://user-images.githubusercontent.com/26887703/176911335-cc9ca8a2-12e8-447e-b84e-3354123fa0b1.png)

**Comment by afa7789** [comment]
@joaomoreno

**Comment by ericbunton** [comment]
> EDIT: I checked if I had the upstream set, which I didn't, whoops, but it seems that it's still the same result.
> 
> I have a setup where I have this folder with a bare-cloned repository and have worktrees for each branches I work on, for some reason it thinks all of my branches are unpublished, which it isn't the case.
> 
> ```
> .git          // gitdir: ./.bare
> .bare/        // bare cloned repository
> staging/
> production/
> development/
> feature-xxx/ 
> bugfix-xxx/
> ```

@intrnl Can you share what you did to verify and set upstream? I feel kinda dumb cause apparently what I've done just isn't correct. I've got a very similar structure and am seeing this same behavior.

**Comment by madorodam** [comment]
There are workarounds for the Dev Container issue, like:

- create a workspace and save it in a folder that contains all git worktrees.
- add the worktree git folder to the workspace.
- create the devcontainer in the folder where the workspace was saved (the create devcontainer configuration action does this).
- and in the git worktree folder edit the .git file and replace the gitdir with a relative folder instead of an absolute one (this is the main trick in fact, the other are only docker mount tricks).

Alternatively I think it might work also to add/mount the main git for the worktrees, in the same relative position, using a docker-compose.yaml file and consume it in devcontrainer config. Did not test this myself yet.

**Comment by aaronsgithub** [comment]
@

> There are workarounds for the Dev Container issue, like:
> 
> * create a workspace and save it in a folder that contains all git worktrees.
> * add the worktree git folder to the workspace.
> * create the devcontainer in the folder where the workspace was saved (the create devcontainer configuration action does this).
> * and in the git worktree folder edit the .git file and replace the gitdir with a relative folder instead of an absolute one (this is the main trick in fact, the other are only docker mount tricks).
> 
> Alternatively I think it might work also to add/mount the main git for the worktrees, in the same relative position, using a docker-compose.yaml file and consume it in devcontrainer config. Did not test this myself yet.

@madorodam (or anyone else familiar), would you be able to share any more details on how to get this working?

Perhaps by sharing an example directory structure?

**Comment by well1791** [comment]
I have an alternative solution that goes like this:

1. clone a bare repo `git clone --single-branch --bare "$repo"`
1. enter the clone and add remote branches
    `git config --local --add remote.origin.fetch "+refs/heads/*:refs/remotes/origin/*"`
1. create 2 worktrees inside
    ```sh
    git worktree add -b rename1 wt/1/$proj_name
    git worktree add -b rename1 wt/1/$proj_name
    ```
1. enter the first worktree `cd wt/1/$proj_name`

here's my fish script for reference

```fish
# file: ~/.config/fish/functions/gcb.fish
function gcb --description 'Clones a repo as bare and set up two worktrees.'
    set repo "$argv[1]"
    set dir "$argv[2]"

    if test -z $dir
        set dir (echo "$repo" | grep -o '[^/]*$')
    end

    set proj_name (string join . (string split . "$dir")[1..-2])

    git clone --single-branch --bare "$repo" "$dir"

    cd "$dir"

    git fetch --all --prune
    git config --local --add remote.origin.fetch "+refs/heads/*:refs/remotes/origin/*"
    git worktree add -b rename1 wt/1/$proj_name
    git worktree add -b rename2 wt/2/$proj_name

    cd wt/1/$proj_name
end
```

In the end I would do something like

```fish
gcb git@github.com:microsoft/vscode.git && code .
```

Hope it helps.

**Comment by moatorres** [comment]
I'm having a similar issue.

**Comment by RalfJung** [comment]
This doesn't just affect dev containers. Handling of rebases is also broken in worktrees.

**Comment by tusharsnx** [comment]
If I'm not wrong, vscode just shows all the worktrees active in the repo, but fails to use the right configuration (`.vscode/`) because there's no way to choose the active worktree being worked on. It's good for submodules but not for worktrees. I'd like to have an option to select the active worktree so that VS Code and extensions pull their configurations from that worktree's directory.

Ideally, VSCode should behave like it's opened in one of the worktrees directories. File explorer, Search, etc., should only work with the currently active worktree. In addition to add, you have an option in the UI to switch to a different worktree. The vscode reloads to open the new worktree.

**Comment by d13** [comment]
In `1.103.0-insider` my SCM Changes view is now spammed with all my worktrees. This is not a good experience by any means. I would expect the changes view to only show me the worktree I'm on, which is how it used to work.

> Edit: using `"git.detectWorktrees": false` for now

<img width="507" height="1135" alt="Image" src="https://github.com/user-attachments/assets/04b5340e-6f18-400b-b0c9-dc6814ee7006" />

<img width="466" height="140" alt="Image" src="https://github.com/user-attachments/assets/5ee38109-aa24-4bea-9bab-2b112d1933aa" />

**Comment by readncode** [comment]
In the 'repositories' view, you can manually select which repositories appear in the 'changes' section (Ctrl+click for multi-select), which already helps reduce clutter. If you select just the main repository, I think you have what you ask for, only the changes of the main repo are visible. The selection persists when you reopen the workspace.

I really appreciate how this feature allows quick switching between the 'changes' view of different worktrees without the need to open them individually.

This feature is quite hidden—I only discovered it after using VS Code for a long time. Not sure how its visibility could be improved, but it's worth surfacing more clearly.

**Comment by lszomoru** [comment]
@d13 thank you very much for your feedback. Adding better support for git worktrees in VS Code has been a long standing feature request and earlier this week we have toggled the `git.detectWorktrees` setting. If you do not want to see the worktees when opening the main git repository you can either toggle the setting, or use the Repositories view as described by @readncode.

We would also appreciate if people that have upvoted/subscribed this feature request over the years would download the latest VS Code Insiders release, try out the feature and provide us with feedback. Thank you very much in advance!

**Comment by d13** [comment]
@lszomoru  - I definitely want worktree support in the SCM. Its been a long-standing issue for me with the SCM statusbar when trying to switch to a branch in a worktree wouldn't work. So much so I created my own private extension to replace it.

My problem is with the experience in the Changes view. @readncode recommendation of having to toggle off each individual worktree means I would have to do that in each worktree's workspace which is lot of annoying and tedious work, especially when you have a lot of worktrees. It would be simpler to have an additional setting to only show the current worktree in the changes view.

**Comment by RalfJung** [comment]
Yeah, agreed -- a worktree should behave entirely as-if it were its own local git repo, except that it shares the object pool with some other git repos. `git` commands generally work solely on the current worktree, and I would expect the same from IDE integration.

**Comment by readncode** [comment]
> having to toggle off each individual worktree means I would have to do that in each worktree's workspace 

From my experience, when I open a worktree in a workspace, only that worktree is visible, the others worktrees appear only in the main repository, so no need to do this for each worktree's workspace.
To hide all other worktrees in the 'Changes' view of the main repo, it is sufficient to click only on the main repo in the repositories view, which deselects all other repos.
Nevertheless, I would also prefer if, by default, only the changes for the main repository were visible, while still having the option to view changes from other repositories by selecting them in the repositories view.

Personally, I like being able to quickly glance at changes from other worktrees without opening each workspace separately, since that can be time-consuming. Sometimes I forget to name my worktrees clearly, so seeing all changes together helps me find what I need more easily. So thank you for this nice feature.

**Comment by lszomoru** [comment]
> @d13 - My problem is with the experience in the Changes view. @readncode recommendation of having to toggle off each individual worktree means I would have to do that in each worktree's workspace which is lot of annoying and tedious work, especially when you have a lot of worktrees. It would be simpler to have an additional setting to only show the current worktree in the changes view.

You can open the each worktree folder either from the Repositories view (right click on a worktree and choose Open In New Window or Open in Current Window) which will give you a workspace scoped to the particular worktree. In this case the Changes view should only show the changes from the current worktree. 


> @RalfJung - Yeah, agreed -- a worktree should behave entirely as-if it were its own local git repo, except that it shares the object pool with some other git repos. git commands generally work solely on the current worktree, and I would expect the same from IDE integration.

That's the goal. Are you seeing any issues when you open a worktree folder in VS Code?

> @readncode, Personally, I like being able to quickly glance at changes from other worktrees without opening each workspace separately, since that can be time-consuming. Sometimes I forget to name my worktrees clearly, so seeing all changes together helps me find what I need more easily. So thank you for this nice feature.

This is the reason why we have opened to show all worktrees when opening the main repository.

Please keep the feedback coming. Thank you!

**Comment by RalfJung** [comment]
> This is the reason why we have opened to show all worktrees when opening the main repository.

Oh I think I see the misunderstanding... it seems like you are treating the "main" directory different from the other worktrees? That doesn't align with my mental model at all -- I just have 3 folders that share the same git object store, none of them is special in any way. It happens to be the case that one contains the `.git` folder and the others just reference it, but that's an implementation detail that shouldn't affect user-visible behavior at all, IMO. When using `git` directly, that's how things work, so it is what I would expect from IDEs by default as well.

But I can live with this being configurable. :)

**Comment by tusharsnx** [comment]
I think an option to open other worktrees in vscode would be a nice feature. If you press the branch button on the status bar, it currently checks out the selected branch in the current directory. It would be better if a new worktree is created for that branch.

For example, when I'm in one of the worktree, I can open other worktrees quickly in vscode with the following command:
 
```sh
code "$(git worktree list | grep "\[my-branch\]" | awk '{print $1}')"
```

**Comment by lszomoru** [comment]
> @tusharsnx, I think an option to open other worktrees in vscode would be a nice feature. If you press the branch button on the status bar, it currently checks out the selected branch in the current directory. It would be better if a new worktree is created for that branch. 

You can use the `Git: Create Worktree..." command from the command palette to create a new worktree for a new/existing branch. Having said that, please feel free to create a new issue for this. Thanks!

**Comment by lszomoru** [comment]
VS Code 1.103 that is available in Insiders and will be released to Stable in the middle of next week will contain the first iteration of our support for  git worktrees. I will go ahead and close this long standing issue. I would like to like to invite everyone to download VS Code 1.103 and try it out and provide us with feedback so that we can continue to improve this experience. Thank you very much in advance.

**Comment by dlschmidt** [comment]
The feature itself looks very promising and the worktree detection seems to be working (just learned about worktrees today, so I can't give complete feedback) However, the actions like "Create Worktree..." are only available over the commands palette but not as shown in the release screenshots, atleast on my side, so the whole "Worktrees" sub menu is missing. 

Also a quick idea: Another command like "Switch worktree" could be implemented, such that after opening a worktree with the existing open command, the user can switch back to the main worktree or a different one.

**Comment by nguyenchristy** [comment]
> The feature itself looks very promising and the worktree detection seems to be working (just learned about worktrees today, so I can't give complete feedback) However, the actions like "Create Worktree..." are only available over the commands palette but not as shown in the release screenshots, atleast on my side, so the whole "Worktrees" sub menu is missing.
> 
> Also a quick idea: Another command like "Switch worktree" could be implemented, such that after opening a worktree with the existing open command, the user can switch back to the main worktree or a different one.

@dlschmidt Thanks for trying it out and the feedback! Did you make sure you were on the 'Repositories' and not the 'Changes'?

<img width="611" height="393" alt="Image" src="https://github.com/user-attachments/assets/1dac15f8-bca0-4587-93ba-0067a915da1d" />

**Comment by dlschmidt** [comment]
@nguyenchristy Okay, my mistake, I was looking at 'Changes' as the entries look so similar to the 'Repository' ones. From a quick glance, all other actions are available for both 'Changes' and 'Repositories', therefore maybe add the actions also to the other one for simplicity?

**Comment by giiyms** [comment]
Is there anyway to have:

./ProjectFolder/
-.git
-Branch1
-Branch2

I set this up using bare for the .git folder:
git clone --bare URL

then create the branches using:
git worktree add branch1 origin/branch1

But when I go into the Branches, they show no changes, just a Publish Branch button:

<img width="375" height="141" alt="Image" src="https://github.com/user-attachments/assets/484f8542-94e4-45eb-b9d1-8cf2702c93a3" />

Now that button works and it does push the changes to the branch, but it will never pull the changes from origin. I think each worktree folder has a .git that has a gitdir pointing to the ../.git but I guess that never get recognised. 

The weird thing is though, I can do a non bare clone, so just:
git clone URL main
then add the worktree up a level:
git worktree add ../branch1 origin/branch1

and that works perfectly, except now I have the .git folder in the /main folder and the main branch checked out.
so this ends up as:
./ProjectFolder/
-Main/.git
-Branch1
-Branch2

**Comment by pickfire** [comment]
Worktree broke dev containers. With the following setup:

- git
  - dev container
- worktree
  - dev container

After creating the worktree, rebuild and reopen in dev container within worktree does not work.

I don't want to create the worktree within the same dev container as I want to separate the port bindings of each worktree within the dev container itself, otherwise I have to manually modify the port number and the services might interfere with each other.

**Comment by alexandertsukanov** [comment]
Hey, folks!  
Is it possible to edit files from different work trees and the main repository using the same window?

Let's say I have the class Foo.java in the core branch. I create Worktree1, and I want to edit Foo.java in Worktree1, not in the core branch. How can I achieve that?

**Comment by nguyenchristy** [comment]
@giiyms You may be creating a worktree in a detached HEAD state. Can you add the[ `-b` flag ](https://git-scm.com/docs/git-worktree) when running your command? For example: `git worktree add -b branch1 Branch1 origin/branch1` to create a local tracking branch? 
If you are still running into problems, can you file a separate issue for this so you can track the progress of it?

**Comment by nguyenchristy** [comment]
> Worktree broke dev containers. With the following setup:
> 
> * git
>   
>   * dev container
> * worktree
>   
>   * dev container
> 
> After creating the worktree, rebuild and reopen in dev container within worktree does not work.
> 
> I don't want to create the worktree within the same dev container as I want to separate the port bindings of each worktree within the dev container itself, otherwise I have to manually modify the port number and the services might interfere with each other.

@pickfire can you file a separate issue for this? Thank you!

**Comment by nguyenchristy** [comment]
@alexandertsukanov  Thank you for trying out worktrees! Currently, we are not supporting adding worktrees to your current workspace. Feel free to file an issue for this for the next milestone as we are still experimenting with worktree support!

> Hey, folks! Is it possible to edit files from different work trees and the main repository using the same window?
> 
> Let's say I have the class Foo.java in the core branch. I create Worktree1, and I want to edit Foo.java in Worktree1, not in the core branch. How can I achieve that?
